File: core/src/main/java/net/m3mobile/core/BroadcastRequester.kt
package net.m3mobile.core

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.core.os.bundleOf

@InternalM3Api
abstract class BroadcastRequester {

    protected abstract val requestAction: String
    protected abstract val typeKey: String
    protected abstract val typeValue: String
    protected open val extras: Bundle = bundleOf()
    protected abstract val context: Context

    fun runBroadcast() {
        val intent = Intent(requestAction)
            .putExtras(extras)
            .putExtra(typeKey, typeValue)
        context.sendBroadcast(intent)
    }
}

File: dotnet/M3Mobile.M3Sdk.Xamarin/Internal/BroadcastRequester.cs
using Android.Content;
using Android.OS;

namespace M3Sdk.Xamarin.Internal
{
    internal abstract class BroadcastRequester
    {
        protected BroadcastRequester(Context context)
        {
            Context = context;
        }

        protected Context Context { get; private set; }

        protected abstract string RequestAction { get; }

        protected virtual string TypeKey
        {
            get { return string.Empty; }
        }

        protected virtual string TypeValue
        {
            get { return string.Empty; }
        }

        protected virtual Bundle CreateExtras()
        {
            return null;
        }

        /// <summary>
        /// Sends the configured Android broadcast request.
        /// </summary>
        internal virtual void Request()
        {
            RunBroadcast();
        }

        protected void RunBroadcast()
        {
            var intent = new Intent(RequestAction);

            var extras = CreateExtras();

            if (extras != null)
                intent.PutExtras(extras);

            if (!string.IsNullOrEmpty(TypeKey))
                intent.PutExtra(TypeKey, TypeValue);

            Context.SendBroadcast(intent);
        }
    }
}
